numpy/core/src/multiarray/einsum.dispatch.c.src

```diff
@@ -0,0 +1,1866 @@
+            data_out += UNROLL_BY;
+        }
+        /* Finish off the loop */
+        goto finish_after_unrolled_loop;
+    }
+#endif
+
+    /* Unroll the loop by UNROLL_BY */
+    while (count >= UNROLL_BY) {
+        count -= UNROLL_BY;
+
+#if NPY_SIMD && @float32@
+/**begin repeat2
+ * #i = 0, 1#
+ */
+        pack_size = @i@ * UNROLL_BY / 2;
+        a = npyv_mul_f32(npyv_load_f32(data0+pack_size), npyv_load_f32(data1+pack_size));
+        b = npyv_add_f32(a, npyv_load_f32(data_out+pack_size));
+        npyv_store_f32(data_out+pack_size, b);
+/**end repeat2**/
+#elif NPY_SIMD_F64 && @float64@
+/**begin repeat2
+ * #i = 0, 1, 2, 3#
+ */
+        pack_size = @i@ * UNROLL_BY / 4;
+        a = npyv_mul_f64(npyv_load_f64(data0+pack_size), npyv_load_f64(data1+pack_size));
+        b = npyv_add_f64(a, npyv_load_f64(data_out+pack_size));
+        npyv_store_f64(data_out+pack_size, b);
+/**end repeat2**/
+#else
+    for (int i = 0; i < UNROLL_BY; i++) {
+        data_out[i] = @to@(@from@(data0[i]) *
+                                 @from@(data1[i]) +
+                                 @from@(data_out[i]));
+    }
+#endif
+        data0 += UNROLL_BY;
+        data1 += UNROLL_BY;
+        data_out += UNROLL_BY;
+    }
+
+    /* Finish off the loop */
+    goto finish_after_unrolled_loop;
+}
+
+/* Some extra specializations for the two operand case */
+static void
+@name@_sum_of_products_stride0_contig_outcontig_two(int nop, char **dataptr,
+                                npy_intp const *NPY_UNUSED(strides), npy_intp count)
+{
+    @temptype@ value0 = @from@(*(@type@ *)dataptr[0]);
+    @type@ *data1 = (@type@ *)dataptr[1];
+    @type@ *data_out = (@type@ *)dataptr[2];
+    npy_intp pack_size;
+#if NPY_SIMD && @float32@
+    npyv_f32 a, b, value0_sse;
+#elif NPY_SIMD_F64 && @float64@
+    npyv_f64 a, b, value0_sse;
+#endif
+
+    NPY_EINSUM_DBG_PRINT1("@name@_sum_of_products_stride0_contig_outcontig_two (%d)\n",
+                                                    (int)count);
+
+/* This is placed before the main loop to make small counts faster */
+finish_after_unrolled_loop:
+    for (; count < UNROLL_BY && count > 0; count--) {
+        data_out[count-1] = @to@(value0 *
+                                 @from@(data1[count-1]) +
+                                 @from@(data_out[count-1]));
+    }
+    if (count == 0) {
+        return;
+    }
+
+#if NPY_SIMD && @float32@
+    value0_sse = npyv_setall_f32(value0);
+
+    /* Use aligned instructions if possible */
+    if (EINSUM_IS_ALIGNED(data1) && EINSUM_IS_ALIGNED(data_out)) {
+        /* Unroll the loop by UNROLL_BY */
+        while (count >= UNROLL_BY) {
+            count -= UNROLL_BY;
+
+/**begin repeat2
+ * #i = 0, 1#
+ */
+            pack_size = @i@ * UNROLL_BY / 2;
+            a = npyv_mul_f32(value0_sse, npyv_loada_f32(data1+pack_size));
+            b = npyv_add_f32(a, npyv_loada_f32(data_out+pack_size));
+            npyv_storea_f32(data_out+pack_size, b);
+/**end repeat2**/
```

Comment:
can you just use a regular `for` loop here? I don't expect the compiler to have any trouble unrolling it, and it makes things easier to read.

Reply:
good point, I will transform to regular `for` loop.